hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java

```diff
@@ -87,27 +87,30 @@ public void scanContainer(Container<?> c)
     ContainerData containerData = c.getContainerData();
     long containerId = containerData.getContainerID();
     logScanStart(containerData);
-    ScanResult result = c.scanData(throttler, canceler);
-
-    // Metrics for skipped containers should not be updated.
-    if (result.getFailureType() == DELETED_CONTAINER) {
-      LOG.error("Container [{}] has been deleted.",
-          containerId, result.getException());
-      return;
-    }
-    if (!result.isHealthy()) {
-      LOG.error("Corruption detected in container [{}]. Marking it UNHEALTHY.",
-          containerId, result.getException());
-      boolean containerMarkedUnhealthy = controller.markContainerUnhealthy(containerId, result);
-      if (containerMarkedUnhealthy) {
-        metrics.incNumUnHealthyContainers();
+    DataScanResult result = c.scanData(throttler, canceler);
+
+    if (result.isDeleted()) {
+      LOG.debug("Container [{}] has been deleted during the data scan.", containerId);
+    } else {
+      if (!result.isHealthy()) {
+        logUnhealthyScanResult(containerId, result, LOG);
+
+        // Only increment the number of unhealthy containers if the container was not already unhealthy.
+        if (controller.markContainerUnhealthy(containerId, result)) {
+          metrics.incNumUnHealthyContainers();
```

Comment:
This method does not check if before the scan the container was healthy, then before incrementing the metric we should check if there is a state transition or if the container was already unhealthy.

Reply:
I think we should log unhealthy results on each iteration, even if the container was unhealthy on a previous iteration. The reason it is unhealthy may have changed.

For the metric, the `controller.markContainerUnhealthy` call returns a boolean indicating if the container state was changed, which we use to increment the metric. Technically you could have an execution like this:
1. Background scanner marks container unhealthy
2. On-demand scanner moves it back to healthy
3. Background scanner increments the metric

I think this is expected behavior though, because the counts of each container in the metric are from the background scanner's point of view. Since it is a slow running process, states of containers it already scanned can change after it has scanned them but before it has scanned them again.

Reply:
I've added a TODO here about how #7335 will change the way this metric is used when we merge it in from master.